stellargraph/layer/knowledge_graph.py

```diff
@@ -137,6 +137,96 @@ def embeddings(model):
 
         return node, rel
 
+    @staticmethod
+    def rank_edges_against_all_nodes(model, test_data, known_edges_graph):
+        """
+        Returns the ranks of the true edges in ``test_data``, when scored against all other similar
+        edges.
+
+        For each input edge ``E = (s, r, o)``, the score of the *modified-object* edge ``(s, r, n)``
+        is computed for every node ``n`` in the graph, and similarly the score of the
+        *modified-subject* edge ``(n, r, o)``.
+
+        This computes "raw" ranks: the score of each edge is ranked against all of the
+        modified-object and modified subject-ones, for instance, if ``E = ("a", "X", "b")`` has
+        score 3.14, and only one modified-object edge has a higher score (e.g. ``("a", "X", "c")``),
+        then the raw modified-object rank for ``E`` will be 2; if all of the ``(n, "X", "b")`` edges
+        have score less than 3.14, then the raw modified-subject rank for ``E`` will be 1.
+
+        Args:
+            model (tensorflow.keras.Model): a Keras model created using a ``ComplEx`` instance.
+
+            test_data: the output of :meth:`KGTripleGenerator.flow` on some test triples
+
+            known_edges_graph (StellarGraph):
+                a graph instance containing all known edges/triples
+
+        Returns:
+            A numpy array of integer raw ranks. It has shape ``N × 2``, where N is the number of
+            test triples in ``test_data``; the first column (``array[:, 0]``) holds the
+            modified-object ranks, and the second (``array[:, 1]``) holds the modified-subject
+            ranks.
+        """
+
+        if not isinstance(test_data, KGTripleSequence):
+            raise TypeError(
+                "test_data: expected KGTripleSequence; found {type(test_data).__name__}"
+            )
+
+        num_nodes = known_edges_graph.number_of_nodes()
+
+        def ranks(pred, true_ilocs, true_is_source):
```

Comment:
I think this function could be pulled out of `ComplEx` and re-used for all our other knowledge graph algorithms. But this function is likely the main experimental blocker so I'm happy for this do be done as part of #1060

Reply:
Similar to #865, I'd prefer to not start abstracting before we're more sure about what's useful, but yes, this is definitely likely to be able to be shared in some form.